pkg/controller/instance/instance_controller.go

```diff
@@ -141,12 +149,40 @@ func (r *Reconciler) Reconcile(request ctrl.Request) (ctrl.Result, error) {
 		return reconcile.Result{}, err // OV not found has to be retried because it can really have been created after Instance
 	}
 
-	// ---------- 2. First check if we should start execution of new plan ----------
+	// ---------- 2. Check if the object is being deleted, start cleanup plan ----------
 
-	planToBeExecuted, err := instance.GetPlanToBeExecuted(ov)
-	if err != nil {
-		return reconcile.Result{}, err
+	var planToBeExecuted *string
+
+	// A delete request is indicated by a non-zero 'metadata.deletionTimestamp',
+	// see https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#finalizers
+	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
+		if _, hasCleanupPlan := ov.Spec.Plans[kudov1beta1.CleanupPlanName]; hasCleanupPlan {
+			if err := r.addFinalizer(instance); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	} else {
+		log.Printf("InstanceController: Instance %s/%s is being deleted", instance.Namespace, instance.Name)
+		if contains(instance.ObjectMeta.Finalizers, instanceCleanupFinalizerName) {
+			// Only start the cleanup plan if it isn't currently running or if
+			// it isn't already completed.
+			lastExecutedPlanStatus := instance.GetLastExecutedPlanStatus()
+			if lastExecutedPlanStatus == nil || lastExecutedPlanStatus.Name != kudov1beta1.CleanupPlanName {
+				cleanupPlan := kudov1beta1.CleanupPlanName
+				planToBeExecuted = &cleanupPlan
```

Comment:
oh wait, now I just went downstairs and realized this might not be enough actually, take a look at what happens in `StartPlanExecution`

Reply:
Not sure what you're referring to. Is it that an existing plan might still be pending and wouldn't get updated to failed? Might be that we need to fail pending plan first.